crates/matrix-sdk/src/event_cache/paginator.rs

```diff
@@ -0,0 +1,564 @@
+    ///
+    /// May return an error if it's already paginating, or if the call to
+    /// /messages failed.
+    pub async fn paginate_backward(&self) -> Result<PaginationResult, PaginatorError> {
+        self.paginate(Direction::Backward).await
+    }
+
+    /// Runs a forward pagination, from the current state of the object.
+    ///
+    /// Will return immediately if we did hit the end of the timeline before.
+    ///
+    /// May return an error if it's already paginating, or if the call to
+    /// /messages failed.
+    pub async fn paginate_forward(&self) -> Result<PaginationResult, PaginatorError> {
+        self.paginate(Direction::Forward).await
+    }
+
+    async fn paginate(&self, dir: Direction) -> Result<PaginationResult, PaginatorError> {
+        self.check_state(PaginatorState::Idle)?;
+
+        let token = match dir {
+            Direction::Backward => {
+                let prev_batch_token = self.prev_batch_token.lock().await;
+                if prev_batch_token.is_none() {
+                    return Ok(PaginationResult { events: Vec::new(), hit_end_of_timeline: true });
+                };
+                prev_batch_token.clone()
+            }
+
+            Direction::Forward => {
+                let next_batch_token = self.next_batch_token.lock().await;
+                if next_batch_token.is_none() {
+                    return Ok(PaginationResult { events: Vec::new(), hit_end_of_timeline: true });
+                };
+                next_batch_token.clone()
+            }
+        };
+
+        // Note: it's possible two callers have checked the state and both figured it's
+        // idle. This check makes sure there's at most one which can set the
+        // state to paginating, preventing a race condition.
+        if self.state.set_if_not_eq(PaginatorState::Paginating).is_none() {
+            return Err(PaginatorError::InvalidPreviousState {
+                expected: PaginatorState::Idle,
+                actual: PaginatorState::Paginating,
+            });
+        }
+
+        let response_result =
+            self.room.messages(MessagesOptions::new(dir).from(token.as_deref())).await;
+
+        // In case of error, reset the state to idle.
+        let response = match response_result {
+            Ok(res) => res,
+            Err(err) => {
+                self.state.set(PaginatorState::Idle);
+                return Err(err);
+            }
+        };
+
+        let hit_end_of_timeline = match dir {
+            Direction::Backward => {
+                let hit = response.end.is_none();
+                *self.prev_batch_token.lock().await = response.end;
+                hit
+            }
+            Direction::Forward => {
+                let hit = response.end.is_none();
+                *self.next_batch_token.lock().await = response.end;
+                hit
+            }
+        };
+
+        // TODO: what to do with state events?
```

Comment:
Some state events appear in the timeline (e.g. location shares) so maybe handle them the same as everything else?

Reply:
(Indeed, and related to the comment about lazy member loading above — deferred to later)